server/src/main/java/org/elasticsearch/search/retriever/rankdoc/RankDocsQuery.java

```diff
@@ -33,142 +40,307 @@
+        if (sources.length > 0 && false == onlyRankDocs) {
+            var bq = new BooleanQuery.Builder();
+            for (var source : sources) {
+                bq.add(source, BooleanClause.Occur.SHOULD);
+            }
+            this.tailQuery = bq.build();
+        } else {
+            this.tailQuery = null;
+        }
+        this.onlyRankDocs = onlyRankDocs;
+    }
+
+    private RankDocsQuery(RankDoc[] docs, Query topQuery, Query tailQuery, boolean onlyRankDocs) {
         this.docs = docs;
-        this.segmentStarts = segmentStarts;
-        this.contextIdentity = contextIdentity;
+        this.topQuery = topQuery;
+        this.tailQuery = tailQuery;
+        this.onlyRankDocs = onlyRankDocs;
     }
 
-    @Override
-    public Query rewrite(IndexSearcher searcher) throws IOException {
-        if (docs.length == 0) {
-            return new MatchNoDocsQuery();
+    private static int binarySearch(RankDoc[] docs, int fromIndex, int toIndex, int key) {
+        return Arrays.binarySearch(docs, fromIndex, toIndex, new RankDoc(key, Float.NaN, -1), Comparator.comparingInt(a -> a.doc));
+    }
+
+    private static int[] findSegmentStarts(IndexReader reader, RankDoc[] docs) {
+        int[] starts = new int[reader.leaves().size() + 1];
+        starts[starts.length - 1] = docs.length;
+        if (starts.length == 2) {
+            return starts;
+        }
+        int resultIndex = 0;
+        for (int i = 1; i < starts.length - 1; i++) {
+            int upper = reader.leaves().get(i).docBase;
+
+            resultIndex = binarySearch(docs, resultIndex, docs.length, upper);
+            if (resultIndex < 0) {
+                resultIndex = -1 - resultIndex;
+            }
+            starts[i] = resultIndex;
         }
-        return this;
+        return starts;
     }
 
     RankDoc[] rankDocs() {
         return docs;
     }
 
     @Override
-    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
-        if (searcher.getIndexReader().getContext().id() != contextIdentity) {
-            throw new IllegalStateException("This RankDocsDocQuery was created by a different reader");
+    public Query rewrite(IndexSearcher searcher) throws IOException {
+        if (tailQuery == null) {
+            return topQuery;
         }
-        return new Weight(this) {
+        boolean hasChanged = false;
+        var topRewrite = topQuery.rewrite(searcher);
+        if (topRewrite != topQuery) {
+            hasChanged = true;
+        }
+        var tailRewrite = tailQuery.rewrite(searcher);
+        if (tailRewrite != tailQuery) {
+            hasChanged = true;
+        }
+        return hasChanged ? new RankDocsQuery(docs, topRewrite, tailRewrite, onlyRankDocs) : this;
+    }
 
+    @Override
+    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+        assert tailQuery != null;
```

Comment:
We should hard fail if tailQuery == null here indicating that a rewrite didn't occur.

Reply:
makes sense, added in 2366d4c0f08d20b5b6537fa2bfc175d9d804e272